src/qutip_qip/noise.py

```diff
@@ -129,18 +129,28 @@ def get_noisy_dynamics(self, dims, pulses, systematic_noise):
             Noisy pulses.
 
         systematic_noise : :class:`.Pulse`
-            The dummy pulse representing pulse independent noise.
+            The dummy pulse representing pulse-independent noise.
         """
-        raise NotImplementedError(
-            "Subclass error needs a method"
-            "`get_noisy_dynamics` to process the noise.")
+        try:
+            # Try to check if the old API is defined.
+            result = self.get_noisy_dynamics(dims, pulses, systematic_noise)
+            warnings.warn(
+                "Using get_noisy_dynamics as the hook function for custom "
+                "noise will be deprecated, "
+                "please use get_noisy_pulses instead.",
+                PendingDeprecationWarning)
+            return result
+        except AttributeError:
+            raise NotImplementedError(
+                "Subclass error needs a method"
+                "`get_noisy_pulses` to process the noise.") from None
```

Comment:
I'm not sure how important this is, but if `.get_noisy_dynamics` exists and raises an `AttributeError` then this code will transform that into a `NotImplementedError` which could be a bit confusing. We could instead do: 
```suggestion
        get_noisy_dynamics = getattr(self, "get_noisy_dynamics", None)
        if get_noisy_dynamics is not None:
            warnings.warn(
                "Using get_noisy_dynamics as the hook function for custom "
                "noise will be deprecated, "
                "please use get_noisy_pulses instead.",
                PendingDeprecationWarning)
            return self.get_noisy_dynamics(dims, pulses, systematic_noise)
        raise NotImplementedError(
            "Subclass error needs a method"
            "`get_noisy_pulses` to process the noise.")
```